src/utils.c

```diff
@@ -275,35 +275,73 @@ get_interval_period_timestamp_units(Interval *interval)
 #endif
 }
 
+#ifdef HAVE_INT64_TIMESTAMP
+#define JAN_3_2000 (2 * USECS_PER_DAY)
+#else
+#define JAN_3_2000 (2 * SECS_PER_DAY)
+#endif
+
+/*
+* The default origin is Monday 2000-01-03. We don't use PG epoch since it starts on a saturday.
+* This makes time-buckets by a week more intuitive and aligns it with
+* date_trunc.
+*/
+#define DEFAULT_ORIGIN (JAN_3_2000)
+#define TIME_BUCKET_TS(period, timestamp, result, shift)			\
+	do															\
+	{															\
+		if (TIMESTAMP_NOT_FINITE(timestamp)) 					\
+			PG_RETURN_TIMESTAMP(timestamp);						\
+		if (period <= 0) \
+			ereport(ERROR, \
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE), \
+				 errmsg("period must be greater then 0"))); \
+		/* shift = shift % period, but use TMODULO */		\
+		TMODULO(shift, result, period);						\
+				 														\
+		if ((shift > 0 && timestamp < DT_NOBEGIN + shift)           \
+			|| shift < 0 && timestamp > DT_NOEND + shift)               \
+			ereport(ERROR,                                          \
+					(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),   \
+					 errmsg("timestamp out of range")));            \
+		timestamp -= shift;											\
+																	\
+		/* result = (timestamp / period) * period */					\
+		TMODULO(timestamp, result, period);								\
+		if (timestamp < 0)												\
+		{																\
+		/*																\
+		 * need to subtract another period if remainder < 0 this only happens \
+		 * if timestamp is negative to begin with and there is a remainder \
+		 * after division. Need to subtract another period since division \
+		 * truncates toward 0 in C99. \
+		 */ 															\
+		result = (result * period) - period;							\
+		}																	\
+		else																\
+			result *= period;												\
+																			\
+		result += shift; \
+	} while (0)
+
+
 TS_FUNCTION_INFO_V1(ts_timestamp_bucket);
 Datum
 ts_timestamp_bucket(PG_FUNCTION_ARGS)
 {
 	Interval   *interval = PG_GETARG_INTERVAL_P(0);
 	Timestamp	timestamp = PG_GETARG_TIMESTAMP(1);
+
+	/*
+	 * USE NARGS and not IS_NULL to differentiate a NULL argument from a call
+	 * with 2 parameters
+	 */
+	Timestamp	origin = (PG_NARGS() > 2 ? PG_GETARG_TIMESTAMPTZ(2) : DEFAULT_ORIGIN);
```

Comment:
Shouldn't this be `PG_GETARG_TIMESTAMP()`? 

Also, even if you need to check `NARGS` here, don't you also need to check for `NULL` value?  It seems weird to do `PG_GETARG_TIMESTAMPTZ(2)` if there is a NULL arg2.

This, b.t.w. also applies to the other arguments (0, 1). In fact, just tested on master and the function segfaults if given a `NULL` interval. I guess this might be fixed in your code given that you added the `STRICT` annotation?

Still, makes me wonder if we generally are bad at checking for NULL args in user-facing functions and we should do some validation/hardening here before 1.0.0?

Reply:
The functions are defined as strict so no need for NULL checking, but I agree we should probably double check for other cases if we handle NULL correctly or functions are defined strict.